src/sync/sync_user.cpp

```diff
@@ -21,44 +21,144 @@
 #include "sync/impl/sync_metadata.hpp"
 #include "sync/sync_manager.hpp"
 #include "sync/sync_session.hpp"
+#include <json.hpp>
 
 namespace realm {
 
+static const std::string b = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";//=
+static std::string base64_decode(const std::string &in) {
+    std::string out;
+
+    std::vector<int> T(256,-1);
+    for (int i=0; i<64; i++) T[b[i]] = i;
+
+    int val=0, valb=-8;
+    for (unsigned char c : in) {
+        if (T[c] == -1) break;
+        val = (val<<6) + T[c];
+        valb += 6;
+        if (valb>=0) {
+            out.push_back(char((val>>valb)&0xFF));
+            valb-=8;
+        }
+    }
+    return out;
+}
+
+static std::vector<std::string> split_token(std::string jwt) {
+    std::string delimiter = ".";
+
+    std::vector<std::string> parts;
+    size_t pos = 0;
+    std::string token;
+
+    while ((pos = jwt.find(delimiter)) != std::string::npos) {
+        token = jwt.substr(0, pos);
+        parts.push_back(token);
+        jwt.erase(0, pos + delimiter.length());
+    }
+
+    parts.push_back(jwt);
+
+    if (parts.size() != 3) {
+        throw SyncError(std::error_code(), "Badly formatted JWT", false);
+    }
+
+    return parts;
+}
+
+RealmJWT::RealmJWT(const std::string& token)
+{
+    auto parts = split_token(token);
+
+    auto second_part = parts[1];
+
+    auto second_part_length = second_part.size();
+    auto extra_chars = second_part_length % 4;
+
+    if (extra_chars != 0) {
+        second_part.insert(second_part_length, second_part_length + 4 - extra_chars, '=');
+    }
+
+    auto json = nlohmann::json::parse(base64_decode(second_part));
+
+    this->token = token;
+    this->expires_at = json["exp"].get<long>();
+    this->issued_at = json["iat"].get<long>();
+    
+    if (json.find("user_data") != json.end()) {
+        this->user_data = json["user_data"].get<std::map<std::string, util::Any>>();
+    }
+}
+
+ObjectSchema SyncUserProfile::schema()
```

Comment:
If I understand this correctly. The user profile is going to be stored in a Realm on device. While it does have some benefits, it also brings about some problems I'm not entirely comfortable with.

- It exposes a Realm lifecycle and threading constraints into API's that might not be suited for it. E.g. the current `SyncUser` does not have threading restrictions.
- Safety issues. Shouldn't the Realm holding these data be encrypted. Where does the encryption key come from?

Reply:
What threading issues does it introduce? SyncUserProfile is a member of SyncUser– they would be stored together.

As far as safety, we did not previously encrypt the user profile in any way. I can ask product and see if this is a requirement.

Reply:
Maybe I misunderstood the code. I thought this turned it into a RealmObject, where instead it was just a C++ data structure?

Previously the SyncUser didn't contain any sensitive data AFAIK, but now it seems that Stitch sends a lot of data that could potentially be a GDPR violation if leaked.

Reply:
It is turned into a realm object, but so is sync user. Maybe I don’t understand why that introduces threading constraints / exposes the realm lifecycle differently than before.

Reply:
Previously, the Sync User was maybe stored in a Realm file, but it was not exposed as a live object. In Java, we deserialized the stored JSON to a static Java Object which meant it could be accessed from any thread.

API's that mix live objects and "normal" objects are pretty painful to work with. At least we discovered this when working on the Permission API's for Java.

There might be enough arguments for exposing the SyncUser as a live object, but e.g. in Java I was planning to keep `SyncUser.createConfiguration()`, which becomes pretty difficult.

I guess we could freeze the SyncUser, but still...

I guess I'm asking what use cases are we solving by turning it into a SyncUser? As I can see it having some unforeseen consequences for how the API's currently work.

I guess I'm askin

Reply:
Using a frozen realm object for SyncUser means that we're keeping the Realm opened (and that version pinned) forever just to avoid copying a small amount of data out, which doesn't seem like a good tradeoff. An unfrozen one means that SyncUser is thread-confined, which probably isn't viable.